### PR TITLE
Adds endpoints for temporary filesets

### PR DESCRIPTION
**Pull request by brendoncarroll**

### CODE REVIEW

File: src/client/pfs/pfs.proto [hunk 1]
```diff
@@ -561,6 +561,16 @@ message DiffFileResponseV2 {
   FileInfo new_file = 2;
 }
 
+message CreateTempFileSetResponse {
```
Comment (brycemcanally) [inline src/client/pfs/pfs.proto:564]: Just a nitpicky thing, but I have been using `tmp` as a shorthand for temporary and it would be nice to be consistent about this in PFS / PPS.

File: src/server/pfs/server/api_server_v2.go [hunk 2]
```diff
@@ -314,3 +318,26 @@ func (a *apiServerV2) DeleteAll(ctx context.Context, request *types.Empty) (resp
...
+	})
+}
+
+func (a *apiServerV2) RenewTempFileSet(ctx context.Context, req *pfs.RenewTempFileSetRequest) (*pfs.CreateTempFileSetResponse, error) {
+	ttl := time.Duration(req.ExtendSeconds) * time.Second
+	t, err := a.driver.renewTempFileSet(ctx, req.FilesetId, ttl)
+	if err != nil {
+		return nil, err
+	}
+	return &pfs.CreateTempFileSetResponse{
```
Comment (brycemcanally) [inline src/server/pfs/server/api_server_v2.go:339]: Is this information useful to return? If so, it seems like it should be in a different message type because this is not the `CreateTempFileSet` rpc.

File: src/server/pfs/server/driver_v2.go [hunk 3]
```diff
@@ -70,7 +72,11 @@ func newDriverV2(
 	if err != nil {
 		return nil, err
 	}
-	chunkStorageOpts := append([]chunk.StorageOption{chunk.WithGarbageCollection(gcClient)}, chunk.ServiceEnvToOptions(env)...)
+	chunkOptsSrv, err := chunk.ServiceEnvToOptions(env)
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:75]: `chunkOptsSrv` -> `chunkStorageOpts` (just reassign below).

File: src/server/pkg/storage/fileset/option.go [hunk 4]
```diff
@@ -52,6 +54,12 @@ func WithMaxOpenFileSets(max int) StorageOption {
 // UWriterOption configures an UnorderedWriter.
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/option.go:54]: I would prefer that we spell this out here, so `UWriterOption` -> `UnorderedWriterOption`. I think we will probably want to minimize the abbreviation of configuration.

File: src/client/pfs_v2.go [hunk 5]
```diff
@@ -40,8 +41,10 @@ func (c APIClient) DeleteFilesV2(repo, commit string, files []string, tag ...str
 // FileOperationClient is not thread safe. Multiple FileOperationClients
 // should be used for concurrent upload.
 type FileOperationClient struct {
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:43]: So I think we should probably tweak this design a bit. I think the basic idea of reusing the `FileOperationV2` functionality makes sense, and I think there is probably more we can refactor on the server side to really make this clean (essentially just have the interpreting of the operations / setting up of the unordered writer in helper functions that both rpcs use). The server side changes can come later, but I think we want to go ahead and tweak this client stuff a bit, because the closing stuff is pretty messy. I think we could set this up such that the `FileOperationV2` and `CreateTempFileSet` client wrappers each have their own abstraction which embeds something that functions similarly to the `FileOperationClient`. The key change though would be that the embedded abstraction would just take a client and be able to send `FileOperationRequestV2`s, but it would have no notion of closing the client. The outer abstraction would have a close method attached to it that would handle the closing differences between the two rpcs. Therefore, we could hand back a client for each rpc that could perform all of the operations as well as have the correct closing semantics.

File: src/client/pfs/pfs.proto [hunk 6]
```diff
@@ -561,6 +561,16 @@ message DiffFileResponseV2 {
...
+  string fileset_id = 1;
+  int64 expires_at = 2;
```
Comment (brycemcanally) [inline src/client/pfs/pfs.proto:566]: After looking over this a bit more, I'm not sure if returning the expire time is that useful and it requires quite a bit of threading through the storage layers. I guess you could use this information to handle renewal for the rare case where a response is really delayed, but I think most use cases will probably just have a fixed renewal based on preexisting knowledge of the default ttl.

File: src/client/pfs_v2.go [hunk 7]
```diff
@@ -267,3 +292,31 @@ func (r *getTarConditionalReader) drain() error {
...
+		return "", err
+	}
+	res, err := foc.closeRecvTempFileSet()
+	if err != nil {
+		return "", err
+	}
+	return res.FilesetId, nil
+}
+
+const TempRepoName = "__tmp__"
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:315]: Since we have a separate endpoint, I don't think this is necessary on the client side. Also, it looks like the below function is unused.

File: src/client/pfs/pfs.proto [hunk 8]
```diff
@@ -561,6 +561,16 @@ message DiffFileResponseV2 {
...
+}
+
+message RenewTempFileSetRequest {
+  string fileset_id = 1;
+  int64 extend_seconds = 2;
```
Comment (brycemcanally) [inline src/client/pfs/pfs.proto:571]: I think you may have mentioned this when we discussed this, but why is this not just interpreted as a duration as compared to number of seconds?

File: src/server/pkg/testpachd/mock_pachd.go [hunk 9]
```diff
@@ -788,6 +796,18 @@ func (api *pfsServerAPI) WalkFileV2(req *pfs.WalkFileRequest, serv pfs.API_WalkF
 	}
 	return errors.Errorf("unhandled pachd mock pfs.WalkFileV2")
 }
+func (api *pfsServerAPI) CreateTempFileSet(srv pfs.API_CreateTempFileSetServer) error {
+	if api.mock.CreateTempFileSet.handler != nil {
+		return api.mock.CreateTempFileSet.handler(srv)
+	}
+	return errors.Errorf("unhandled pachd mock pfs.WalkFileV2")
```
Comment (brycemcanally) [inline src/server/pkg/testpachd/mock_pachd.go:803]: Wrong rpc.

File: src/server/pfs/server/driver_v2.go [hunk 10]
```diff
@@ -961,3 +970,45 @@ func (d *driverV2) deleteCommit(txnCtx *txnenv.TransactionContext, userCommit *p
...
+	// if err := d.txnEnv.WithWriteContext(ctx, func(txnCtx *transactionenv.TransactionContext) error {
+	// 	return d.finishCommitV2(txnCtx, commit, "")
+	// }); err != nil {
+	// 	return "", nil, err
+	// }
+	return commit.ID, expiresAt, nil
+}
+
+func (d *driverV2) renewTempFileSet(ctx context.Context, id string, ttl time.Duration) (*time.Time, error) {
+	if ttl > maxTTL {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:1009]: It might be better to surface this as an error.

File: src/server/pkg/storage/chunk/storage.go [hunk 11]
```diff
@@ -64,15 +69,42 @@ func (s *Storage) CreateSemanticReference(ctx context.Context, name string, chun
 	return s.gcClient.CreateReference(ctx, semanticReference(name, chunk.Hash))
 }
 
+// CreateSemanticReference creates a semantic reference to a chunk.
+func (s *Storage) CreateTemporaryReference(ctx context.Context, name string, chunk *Chunk, ttl time.Duration) (*time.Time, error) {
```
Comment (brycemcanally) [inline src/server/pkg/storage/chunk/storage.go:73]: Why are we passing around pointers to a `time.Time`? The Go recommendation is to pass by value for this type: https://golang.org/pkg/time/#Time

File: src/server/pkg/storage/chunk/storage.go [hunk 12]
```diff
@@ -39,7 +44,7 @@ func (s *Storage) NewReader(ctx context.Context, dataRefs ...*DataRef) *Reader {
 // Chunks are created based on the content, then hashed and deduplicated/uploaded to
 // object storage.
 func (s *Storage) NewWriter(ctx context.Context, tmpID string, f WriterFunc, opts ...WriterOption) *Writer {
-	return newWriter(ctx, s.objClient, s.gcClient, tmpID, f, opts...)
+	return newWriter(ctx, s.objClient, s.gcClient, tmpID, f, append(opts, WithChunkTTL(defaultChunkTTL))...)
```
Comment (brycemcanally) [inline src/server/pkg/storage/chunk/storage.go:47]: I think the order is wrong here. You want the default option first, then the client supplied options afterwards.

File: src/server/pfs/server/driver_v2.go [hunk 13]
```diff
@@ -961,3 +970,45 @@ func (d *driverV2) deleteCommit(txnCtx *txnenv.TransactionContext, userCommit *p
 
 	return nil
 }
+
+func (d *driverV2) createTempFileSet(server pfs.API_CreateTempFileSetServer) (string, *time.Time, error) {
+	ctx := server.Context()
+	opts := []fileset.UWriterOption{fileset.WithWriterOption(fileset.WithTTL(30 * time.Minute))}
+	id := uuid.NewWithoutDashes()
+	var commit *pfs.Commit
+	// TODO: within commit?
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:979]: I think we wanted it in a commit so that the rpcs that access files would work out of the box. Is the concern around the cleanup of the commit? I guess that would not be captured in this model. I think we are eventually going to have a background process that cleans up commits for data retention purposes and maybe it could be what catches the commits associated with cleaned up temporary filesets?

File: src/client/pfs/pfs.proto (hunk 6 above)
Comment (brendoncarroll) [reply src/client/pfs/pfs.proto:566]: I like the lease renewal algorithm that works as follows: Get a lease, check when it expires, schedule a renewal halfway to the expire time, repeat.  This allows the client to know if it has a lease that it can't work with (because it expires too quickly).

File: src/client/pfs/pfs.proto (hunk 8 above)
Comment (brendoncarroll) [reply src/client/pfs/pfs.proto:571]: Do you mean like using the well-known-type for duration? Depending on how protoc is setup those can be a pain, sometimes requiring another conversion in the code.  We don't really need sub-second resolution for leases, since they will typically be on the order of minutes or at lease tens of seconds.  If we use the duration types elsewhere, I can switch to those.

File: src/client/pfs_v2.go (hunk 7 above)
Comment (brendoncarroll) [reply src/client/pfs_v2.go:315]: This is used to construct the commit for the File methods. Instead of having a separate set of functions for commits, we have a sentinel repo that causes the commit id to be interpreted as a temp fileset id.

File: src/client/pfs/pfs.proto [hunk 14]
```diff
@@ -652,6 +661,10 @@ service API {
   rpc InspectFileV2(InspectFileRequest) returns (FileInfo) {}
   // WalkFileV2 walks over all the files under a directory, including children of children.
   rpc WalkFileV2(WalkFileRequest) returns (stream FileInfo) {}
+  // CreateTempFileSet creates a new temp fileset
```
Comment (brycemcanally) [inline src/client/pfs/pfs.proto:664]: These comments need to be updated.

File: src/client/pfs_v2.go [hunk 15]
```diff
@@ -58,11 +58,69 @@ func (c APIClient) NewFileOperationClientV2(repo, commit string) (_ *FileOperati
...
+	})
+}
+
+// CreateTmpFileSetClient is used to create a temporary fileset.
+type CreateTmpFileSetClient struct {
+	client pfs.API_CreateTmpFileSetClient
+	fileOperationCore
+}
+
+func (c APIClient) NewCreateTmpFileSetClient() (ret *CreateTmpFileSetClient, retErr error) {
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:83]: `ret` -> `_`

File: src/client/pfs_v2.go [hunk 16]
```diff
@@ -58,11 +58,69 @@ func (c APIClient) NewFileOperationClientV2(repo, commit string) (_ *FileOperati
...
+	return &CreateTmpFileSetClient{
+		client: client,
+		fileOperationCore: fileOperationCore{
+			client: client,
+		},
+	}, nil
+}
+
+// Close closes the FileOperationClient.
+func (foc *CreateTmpFileSetClient) CloseAndRecv() (*pfs.CreateTmpFileSetResponse, error) {
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:100]: I think `Close` is actually a better name for this method. The gRPC aspect should stay encapsulated in my opinion. Also, update the comment appropriately.

Comment (brycemcanally) [reply src/client/pfs_v2.go:100]: Also `foc` is not really the appropriate variable name. The name for this abstraction might be a bit too wordy.

File: src/server/pfs/server/api_server.go [hunk 17]
```diff
@@ -638,6 +638,16 @@ func (a *apiServer) WalkFileV2(req *pfs.WalkFileRequest, server pfs.API_WalkFile
 	return errV2NotImplemented
 }
 
+// CreateTempFileSet not implemented
```
Comment (brycemcanally) [inline src/server/pfs/server/api_server.go:641]: Update comments.

File: src/client/pfs_v2.go [hunk 18]
```diff
@@ -267,3 +317,13 @@ func (r *getTarConditionalReader) drain() error {
 		}
 	}
 }
+
+const TempRepoName = "__tmp__"
+
+// TmpFileSetCommit
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:323]: Comment?

File: src/server/pfs/server/driver_v2.go [hunk 19]
```diff
@@ -165,10 +171,13 @@ func (d *driverV2) withUnorderedWriter(ctx context.Context, repo, commit string,
 	if err := fs.Close(); err != nil {
 		return err
 	}
-	return d.compactionQueue.RunTaskBlock(ctx, func(m *work.Master) error {
+	if err := d.compactionQueue.RunTaskBlock(ctx, func(m *work.Master) error {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:174]: Just return here.

File: src/client/pfs_v2.go (hunk 7 above)
Comment (brycemcanally) [reply src/client/pfs_v2.go:321]: `TempRepoName` -> `TmpRepoName`

File: src/server/pfs/server/driver_v2.go (hunk 3 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_v2.go:75]: Bump

File: src/server/pfs/server/driver_v2.go [hunk 20]
```diff
@@ -961,3 +970,45 @@ func (d *driverV2) deleteCommit(txnCtx *txnenv.TransactionContext, userCommit *p
...
+	if len(id) != 32 {
+		return errors.Errorf("invalid id (%s)", id)
+	}
+	p := path.Join(tmpRepo, id)
+	_, err := d.storage.RenewFileSet(ctx, p, ttl)
+	return err
+}
+
+func (d *driverV2) inspectCommit(pachClient *client.APIClient, commit *pfs.Commit, blockState pfs.CommitState) (*pfs.CommitInfo, error) {
+	if commit.GetRepo().GetName() == tmpRepo {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:1005]: What is going on here? Why do we return a commit info, shouldn't we just error?

File: src/server/pkg/storage/gc/client.go [hunk 21]
```diff
@@ -24,10 +33,12 @@ var (
 //  * Source - the source of the reference, this may be a temporary id, chunk id, or a
 //    semantic name.
 //  * Chunk - the target chunk being referenced.
+//  * ExpiresAt - the time after which a temporary chunk will be destroyed
 type Reference struct {
-	Sourcetype string
+	Sourcetype SourceType
 	Source     string
 	Chunk      string
+	ExpiresAt  *time.Time
```
Comment (brycemcanally) [inline src/server/pkg/storage/gc/client.go:41]: Why is this a pointer?

Comment (brycemcanally) [reply src/server/pkg/storage/gc/client.go:41]: It looks like gorm is operating on a pointer in some queries and a struct in others.

File: src/server/pfs/server/driver_v2.go (hunk 20 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_v2.go:1002]: We want to call inspect commit on all the accessors before proceeding.  This creates a phony commit info so the accessors can proceed as usually without any changes.

File: src/client/pfs_v2.go [hunk 22]
```diff
@@ -58,11 +58,69 @@ func (c APIClient) NewFileOperationClientV2(repo, commit string) (_ *FileOperati
...
+	}
+	return &CreateTmpFileSetClient{
+		client: client,
+		fileOperationCore: fileOperationCore{
+			client: client,
+		},
+	}, nil
+}
+
+// Close closes the FileOperationClient.
```
Comment (brycemcanally) [inline src/client/pfs_v2.go:99]: `FileOperationClient` -> `CreateTmpFileSetClient`

File: src/server/pkg/storage/fileset/storage.go [hunk 23]
```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
...
+		defer cancel()
+		return cb(ctx2)
+	})
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+	return s.Delete(ctx, prefix)
+}
+
+func (s *Storage) GetExpiresAt(ctx context.Context, prefix string) (time.Time, error) {
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/storage.go:380]: Is this dead code? I cannot find a call site?

File: src/server/pkg/storage/fileset/storage.go [hunk 24]
```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
...
+	if expiresAt == nil {
+		return time.Time{}, ErrNoFileSetFound
+	}
+	return *expiresAt, nil
+}
+
+// RenewDuring ensures that none of the filesets under prefix expire by repeatedly calling SetTTL in a separate goroutine
+// until cb returns.
+// RenewDuring attempts to delete the contents of prefix after cb has returned
+func (s *Storage) RenewDuring(ctx context.Context, prefix string, cb func(context.Context) error) error {
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/storage.go:348]: I don't think this model is going to work for us. I think we need to be able to associate one `expiresAt` with the prefix rather than one for each individual path under that prefix. When we move to postgres, we will be cleaning up the semantic path, so I can imagine a scenario where a path under the prefix expires and is deleted, but we don't catch that it was deleted during the renewal since the renewal just looks at the existing paths. Or, we could just be explicit about which paths to renew, but I kind of like the model where we are renewing on a prefix.

File: src/server/pkg/storage/fileset/storage.go [hunk 25]
```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
 	})
 }
 
+// SetTTL sets the time-to-live for the prefix. If the prefix contains multiple filesets
+// the ttl will be set on all of them and the soonest expiration time will be returned.
+// if no fileset is found SetTTL returns ErrNoFileSetFound
+func (s *Storage) SetTTL(ctx context.Context, prefix string, ttl time.Duration) (time.Time, error) {
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/storage.go:324]: I think we should be a bit more consistent with the naming conventions for this stuff. So, if we need a function like this (I think we may not if we implement renewal on a prefix), then something like `RenewFileSet` seems more consistent and the last parameter would be `expiresAt`. Also, the scoped function below could be `WithRenewFileSet`.

File: src/server/pkg/storage/fileset/storage.go [hunk 26]
```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
...
+	}
+	return *expiresAt, nil
+}
+
+// RenewDuring ensures that none of the filesets under prefix expire by repeatedly calling SetTTL in a separate goroutine
+// until cb returns.
+// RenewDuring attempts to delete the contents of prefix after cb has returned
+func (s *Storage) RenewDuring(ctx context.Context, prefix string, cb func(context.Context) error) error {
+	const defaultTTL = 10 * time.Minute
+	eg, ctx2 := errgroup.WithContext(ctx)
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/storage.go:350]: I tend to name derived contexts with this pattern `*Ctx`. So, this could be `errCtx`. I think that is a bit more informative.

File: src/server/pkg/storage/fileset/storage.go [hunk 27]
```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
...
+	ctx2, cancel := context.WithCancel(ctx2)
+	eg.Go(func() error {
+		var expiresAt time.Time
+		var err error
+		for {
+			expiresAt, err = s.SetTTL(ctx2, prefix, defaultTTL)
```
Comment (brycemcanally) [inline src/server/pkg/storage/fileset/storage.go:356]: I don't think this is the algorithm we want for two reasons: it is not as simple as renewing at a fixed period, and it will actually result with more requests when high latency is encountered which is probably not the appropriate behavior for the system as a whole. I think we should look at this as more of a cooperative process, and error out appropriately when high latency is encountered (so a renewal will eventually fail in this case).

File: src/server/pfs/server/driver_v2.go [hunk 28]
```diff
@@ -144,53 +143,52 @@ func (d *driverV2) getSubFileSet() int64 {
...
+}
 
-	defer func() {
-		if err := d.storage.Delete(ctx, path.Join(tmpPrefix, subFileSetPath)); retErr == nil {
-			retErr = err
-		}
-	}()
-	if err := cb(fs); err != nil {
-		return err
+func (d *driverV2) withTempUnorderedWriter(ctx context.Context, compact bool, cb func(*fileset.UnorderedWriter) error) (string, error) {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:159]: `Temp` -> `Tmp`

File: src/server/pfs/server/driver_v2.go [hunk 29]
```diff
@@ -144,53 +143,52 @@ func (d *driverV2) getSubFileSet() int64 {
...
 	n := d.getSubFileSet()
 	subFileSetStr := fileset.SubFileSetStr(n)
 	subFileSetPath := path.Join(repo, commit, subFileSetStr)
-	fs, err := d.storage.New(ctx, path.Join(tmpPrefix, subFileSetPath), subFileSetStr)
+	id, err := d.withTempUnorderedWriter(ctx, false, cb)
 	if err != nil {
 		return err
 	}
+	tmpPath := path.Join(tmpRepo, id)
+	return d.storage.Copy(ctx, tmpPath, subFileSetPath, defaultTTL)
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:156]: Why do the output filesets have a ttl? The output filesets should be permanent.

**Comment by brycemcanally** [comment]
Take another pass over the expires at code, particularly the code for propagating it back up the layers. I think some of it is dead code after the removal of the expires at from the grpc response.

File: src/server/pkg/storage/gc/client.go (hunk 21 above)
Comment (brendoncarroll) [reply src/server/pkg/storage/gc/client.go:41]: In our model it's optional. When interacting with the database we use `sql.NullTime`

File: src/server/pkg/storage/fileset/storage.go (hunk 23 above)
Comment (brendoncarroll) [reply src/server/pkg/storage/fileset/storage.go:380]: Not used, but I think it's useful to expose this.  The API is pretty straightforward. I think it may be a good idea to have tests to ensure commits do not have a TTL associated with them.  Maybe even do that as part of a set of live integrity checks.

File: src/server/pkg/storage/fileset/storage.go (hunk 24 above)
Comment (brendoncarroll) [reply src/server/pkg/storage/fileset/storage.go]: This has changed to take a list of paths, each of which will be passed to `SetTTL`

File: src/server/pkg/storage/fileset/storage.go (hunk 25 above)
Comment (brendoncarroll) [reply src/server/pkg/storage/fileset/storage.go]: We discussed this offline. `RenewDuring` -> `WithRenewal` to be consistent using `With*` for callback style APIs.  `SetTTL` is more literally what's going on, and could be used to set a TTL on a key without one.  After postgres this will no longer talk to the garbage collector client.

File: src/server/pkg/storage/fileset/storage.go (hunk 27 above)
Comment (brendoncarroll) [reply src/server/pkg/storage/fileset/storage.go]: This is changed to a simple ticker. And due to the new semantics of `WithRenewal` we now detect slowness with not exist errors.

File: src/server/pfs/server/driver_v2.go (hunk 29 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_v2.go:156]: Oops, that would be bad.

File: src/server/pfs/server/driver_v2.go [hunk 30]
```diff
@@ -28,15 +27,16 @@ import (
...
 const (
-	// tmpPrefix is for temporary storage paths.
-	// TODO Temporary prefix cleanup needs some work.
-	// Paths should get cleaned up in the background.
-	tmpPrefix            = "tmp"
 	storageTaskNamespace = "storage"
+	tmpRepo              = client.TmpRepoName
+	maxTTL               = 30 * time.Minute
+	defaultTTL           = 10 * time.Minute
+	driftMargin          = defaultTTL / 4
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:39]: This seems to be unused.

File: src/server/pkg/storage/fileset/storage.go (hunk 23 above)
Comment (brycemcanally) [reply src/server/pkg/storage/fileset/storage.go:395]: I think we should remove it for the time being, because it would need to be changed anyway to accurately reflect the new model. I think most people would interpret this function as returning the expires at for a prefix, which is not how the garbage collection expires at works. Since it operates on independent paths, some paths may be removed much later than the returned time (and some paths might not even be removed at all if they don't have an expires at). I agree that integrity checks will be useful, but I don't think we should keep this around as is just because it could technically work to check if a commit has an expires at in any subfileset.

File: src/server/pfs/server/driver_v2.go [hunk 31]
```diff
@@ -144,53 +144,60 @@ func (d *driverV2) getSubFileSet() int64 {
...
 	subFileSetStr := fileset.SubFileSetStr(n)
 	subFileSetPath := path.Join(repo, commit, subFileSetStr)
-	fs, err := d.storage.New(ctx, path.Join(tmpPrefix, subFileSetPath), subFileSetStr)
+	id, pathsWritten, err := d.withTmpUnorderedWriter(ctx, false, cb)
 	if err != nil {
 		return err
 	}
+	// collect paths for renewal
+	tmpPath := path.Join(tmpRepo, id)
+	return d.storage.WithRenewal(ctx, defaultTTL, pathsWritten, func(ctx context.Context) error {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_v2.go:158]: I don't think this is going to accomplish what we want. A long upload could still result with temporary subfilesets getting deleted since we wouldn't renew until the upload is finished. I was thinking about this a bit, and I think a model we could use that would work across the board would be to have a `Renewer` abstraction that could be passed around and you would be able to make calls on it to kick off renewals for new paths. Then you could have a `Close` function on it that would cancel all of the renewals that have been setup. This means you could setup the `Renewer` in the outer scope and defer the close to be sure that all the renewal goroutines are cleaned up (it might even make sense to just have a `WithRenewer` function). So, how this would work in the unordered writer case is that a `Renewer` would be setup in the outer scope (`withCommitWriter` in this case) and the unordered writer would take the `Renewer` as a parameter and add new paths to renew as it serializes subfilesets. Then when the outer scope is finished, all the renewals that were setup are canceled. For compaction the parent node would setup a `Renewer` for the child nodes and pass it in. Then when the children finish and the parent compacts them, the `Renewer` would be closed which would cancel all of the renewals.

**Review by brycemcanally — APPROVED** [review]
lgtm
